docs/proposals/20200511-clusterctl-extensible-template-processing.md

```diff
@@ -0,0 +1,316 @@
+to explore templating tools to minimize duplication and assert type
+correctness.
+
+
+Users looking to deploy workload clusters may be overwhelmed by the number of
+variables needed to configure a workload cluster. Some of these options may be
+circumvented via defaults provided by the operators as part of the templating
+tool.
+
+
+This proposal aims to extend clusterctl’s current templating mechanism of
+simple variable substitution to support other templating formats natively,
+thus benefiting the operators to deploy workload clusters with minimal
+configuration.
+
+### Goals
+
+
+- Clusterctl will not be changing the way it gets values for yaml variables.
+  That is, it will get values from environment variables and the config file.
+- To provide an interface such that other templating solutions may be
+  supported.
+- To implement ytt as the first templating extension.
+
+### Non-Goals
+- To replace existing mechanisms such as kustomize.
+- To replace Clusterctl default mode (environment variables substitution
+  templating).
+
+### Future Work
+- To automatically detect template files for the appropriate template engine.
+- To use the configuration file to determine which provider should use
+  prefered templating mechanisms. For example,
+ ```yaml
+  yamlProcessors:
+  - provider: aws
+    processor: ytt
+ ```
+  We will be leaning towards using a flag for the initial implementation.
+
+## Proposal
+
+This section outlines the high level work that needs to be done in order to
+accomplish the task of supporting other templating tools.
+
+### User Stories
+
+
+- As a clusterctl user, I would like to preserve the current behavior of using
+  existing workload cluster templates with simple variable substitution from
+  the clusterctl config and environment variables.
+- As an infrastructure provider, I would like to provide templated yaml (e.g.
+  ytt templates) with defaults and have that be consumed by
+  `clusterctl config cluster` to generate cluster manifests.
+- As an infrastructure provider, I would like to provide a set of ytt
+  templates in a compressed file format such as zip, gzip or tar containing
+  all the ytt template artifacts and have that be consumed by `clusterctl
+  config cluster` to generate cluster manifests.
+- As a developer, I would like to easily provide a way to extend clusterctl
+  with my preferred templating solution.
+
+
+### Implementation Details/Notes/Constraints
+
+
+Currently the clusterctl command fetches yaml for the provider components
+(yaml defining the management cluster components) and the workload cluster
+templates.
+
+
+#### UX for using Extensible Templating Tools
+We can use flags to activate the use of a specific templating tool as follows,
+```
+# Retrieve variables required by the ytt templates
+clusterctl config cluster <cluster-name> --infrastructure <provider-name> --list-variables --experimental-template-engine=ytt
```

Comment:
Probably off-topic, `list-variables` feels like it should be a sub-command

Reply:
it does, I was thinking the same thing above `clusterctl init --list-images` yesterday... and they both require credentials to be set which seems wrong since nothing is actually getting deployed (but yeah definitely off topic)

Reply:
off topic but @CecileRobertMichon now you don't have to require credentials for that command. https://github.com/kubernetes-sigs/cluster-api/pull/3059